module/zfs/dmu_send.c

```diff
@@ -2482,11 +2482,15 @@ receive_object(struct receive_writer_arg *rwa, struct drr_object *drro,
 	} else if (drro->drr_type != doi.doi_type ||
 	    drro->drr_blksz != doi.doi_data_block_size ||
 	    drro->drr_bonustype != doi.doi_bonus_type ||
-	    drro->drr_bonuslen != doi.doi_bonus_size) {
+	    drro->drr_bonuslen != doi.doi_bonus_size ||
+	    (!(drro->drr_dn_slots == 0 &&
+	    doi.doi_dnodesize == DNODE_MIN_SIZE) &&
```

Comment:
Is the `!(drro->drr_dn_slots == 0 && doi.doi_dnodesize == DNODE_MIN_SIZE)` portion of this conditional needed at all?  It seems like it's sufficient to solely rely on a difference in the slot count.  If they don't match then worst case we reclaim and reallocate it based on the size provided by the send stream.  If we do need to keep this check for some reason we should find a way to document why since the reason isn't clear (to me at least) with the code alone.

Reply:
I included that portion because send streams from ZFS versions that don't know about large dnodes will have `drro->drr_dn_slots == 0`, which is treated as if it were set to 1. I wanted to avoid reclaiming the object in this case, mainly because I'm not sure what side effects that will have. But if it's safe to reclaim and reallocate the object then we could certainly drop that part.

Reply:
Good thought, my reading of the code is that this is OK and the side effects will be limited too the following all of which are safe.  So this LGTM.

* Kicking any unreferenced dbufs for the dnode out of the cache which is safe.
* Optimizing the bonus space used by SAs.
* A little additional IO since the dnode will be dirtied.